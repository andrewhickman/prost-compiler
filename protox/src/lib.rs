//! A rust implementation of the protobuf compiler.
//!
//! For convenient compilation of protobuf source files in a single function, see
//! [`compile()`]. For more options see [`Compiler`].
//!
//! # Examples
//!
//! Usage with prost-build:
//!
//! ```
//! # use std::{env, fs, path::PathBuf};
//! # use prost::Message;
//! # let tempdir = tempfile::TempDir::new().unwrap();
//! # env::set_current_dir(&tempdir).unwrap();
//! # env::set_var("OUT_DIR", tempdir.path());
//! # fs::write("root.proto", "").unwrap();
//! let file_descriptors = protox::compile(["root.proto"], ["."]).unwrap();
//! prost_build::compile_fds(file_descriptors).unwrap();
//! ```
//!
//! For better error messages, enable the `fancy` feature of `miette` and return a
//! [`miette::Result`](https://docs.rs/miette/latest/miette/type.Result.html) from your `main()` function.
#![warn(missing_debug_implementations, missing_docs)]
#![deny(unsafe_code)]
#![doc(html_root_url = "https://docs.rs/protox/0.3.3/")]

pub mod file;

mod compile;
mod error;

use std::path::Path;

pub use self::compile::Compiler;
pub use self::error::Error;

/// Compiles a set of protobuf files using the given include paths.
///
/// For more control over how files are compiled, see [`Compiler`]. This function is equivalent to:
///
/// ```rust
/// # use protox::Compiler;
/// # fn main() -> Result<(), protox::Error> {
/// # let files: Vec<std::path::PathBuf> = vec![];
/// # let includes: Vec<std::path::PathBuf> = vec![".".into()];
/// let file_descriptor_set = Compiler::new(includes)?
///     .include_source_info(true)
///     .include_imports(true)
///     .open_files(files)?
///     .file_descriptor_set();
/// # Ok(())
/// # }
/// ```
///
/// # Examples
///
/// ```
/// # use std::fs;
/// # use prost_types::{
/// #    DescriptorProto, FieldDescriptorProto, field_descriptor_proto::{Label, Type}, FileDescriptorSet, FileDescriptorProto,
/// #    SourceCodeInfo, source_code_info::Location
/// # };
/// # use protox::compile;
/// # let tempdir = tempfile::TempDir::new().unwrap();
/// # std::env::set_current_dir(&tempdir).unwrap();
/// #
/// fs::write("bar.proto", "
///     message Bar { }
/// ").unwrap();
/// fs::write("root.proto", "
///     import 'bar.proto';
///
///     message Foo {
///         optional Bar bar = 1;
///     }
/// ").unwrap();
///
/// assert_eq!(compile(["root.proto"], ["."]).unwrap(), FileDescriptorSet {
///     file: vec![
///         FileDescriptorProto {
///             name: Some("bar.proto".to_owned()),
///             message_type: vec![DescriptorProto {
///                 name: Some("Bar".to_owned()),
///                 ..Default::default()
///             }],
///             source_code_info: Some(SourceCodeInfo {
///                 /* ... */
/// #               location: vec![
/// #                   Location { path: vec![], span: vec![1, 4, 19], ..Default::default() },
/// #                   Location { path: vec![4, 0], span: vec![1, 4, 19], ..Default::default() },
/// #                   Location { path: vec![4, 0, 1], span: vec![1, 12, 15], ..Default::default() },
/// #               ],
///             }),
///             ..Default::default()
///         },
///         FileDescriptorProto {
///             name: Some("root.proto".to_owned()),
///             dependency: vec!["bar.proto".to_owned()],
///             message_type: vec![DescriptorProto {
///                 name: Some("Foo".to_owned()),
///                 field: vec![FieldDescriptorProto {
///                     name: Some("bar".to_owned()),
///                     number: Some(1),
///                     label: Some(Label::Optional as _),
///                     r#type: Some(Type::Message as _),
///                     type_name: Some(".Bar".to_owned()),
///                     json_name: Some("bar".to_owned()),
///                     ..Default::default()
///                 }],
///                 ..Default::default()
///             }],
///             source_code_info: Some(SourceCodeInfo {
///                 /* ... */
/// #               location: vec![
/// #                   Location { path: vec![], span: vec![1, 4, 5, 5], ..Default::default() },
/// #                   Location { path: vec![3, 0], span: vec![1, 4, 23], ..Default::default() },
/// #                   Location { path: vec![4, 0], span: vec![3, 4, 5, 5], ..Default::default() },
/// #                   Location { path: vec![4, 0, 1], span: vec![3, 12, 15], ..Default::default() },
/// #                   Location { path: vec![4, 0, 2, 0], span: vec![4, 8, 29], ..Default::default() },
/// #                   Location { path: vec![4, 0, 2, 0, 1], span: vec![4, 21, 24], ..Default::default() },
/// #                   Location { path: vec![4, 0, 2, 0, 3], span: vec![4, 27, 28], ..Default::default() },
/// #                   Location { path: vec![4, 0, 2, 0, 4], span: vec![4, 8, 16], ..Default::default() },
/// #                   Location { path: vec![4, 0, 2, 0, 6], span: vec![4, 17, 20], ..Default::default() },
/// #               ],
///             }),
///             ..Default::default()
///         },
///     ],
///     ..Default::default()
/// });
/// ```
pub fn compile(
    files: impl IntoIterator<Item = impl AsRef<Path>>,
    includes: impl IntoIterator<Item = impl AsRef<Path>>,
) -> Result<prost_types::FileDescriptorSet, Error> {
    Ok(Compiler::new(includes)?
        .include_source_info(true)
        .include_imports(true)
        .open_files(files)?
        .file_descriptor_set())
}
