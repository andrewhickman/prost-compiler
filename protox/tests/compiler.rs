use std::fs;

use insta::assert_yaml_snapshot;
use miette::{Diagnostic, JSONReportHandler};
use prost::Message;
use prost_reflect::{DescriptorPool, Value};
use prost_types::{
    source_code_info::Location, FileDescriptorProto, FileDescriptorSet, SourceCodeInfo,
};
use protox::{
    file::{ChainFileResolver, DescriptorSetFileResolver, File, FileResolver, GoogleFileResolver},
    Compiler, Error,
};

struct TestFileResolver {
    files: &'static [(&'static str, &'static str)],
}

impl FileResolver for TestFileResolver {
    fn open_file(&self, name: &str) -> Result<File, Error> {
        if name == "customerror.proto" {
            return Err(Error::new("failed to load file!"));
        }

        for file in self.files {
            if file.0 == name {
                return File::from_source(name, file.1);
            }
        }

        Err(Error::file_not_found(name))
    }
}

fn check(files: &'static [(&'static str, &'static str)]) -> Result<Compiler, Error> {
    let tempdir = tempfile::tempdir().unwrap();
    for (file, source) in files {
        fs::write(tempdir.path().join(file), source).unwrap();
    }

    let mut compiler = Compiler::with_file_resolver(TestFileResolver { files });
    for (file, _) in &files[..files.len() - 1] {
        compiler.open_file(file).unwrap();
    }

    compiler.open_file(files[files.len() - 1].0)?;
    Ok(compiler)
}

fn check_err(files: &'static [(&'static str, &'static str)]) -> serde_json::Value {
    error_to_json(&check(files).unwrap_err())
}

fn error_to_json(err: &dyn Diagnostic) -> serde_json::Value {
    let mut json = String::new();
    JSONReportHandler::new()
        .render_report(&mut json, err)
        .unwrap();
    serde_json::from_str(&json).unwrap()
}

#[test]
fn import_not_found() {
    assert_yaml_snapshot!(check_err(&[("root.proto", "import 'notfound.proto';")]));
}

#[test]
fn import_error() {
    assert_yaml_snapshot!(check_err(&[("root.proto", "import 'customerror.proto';")]));
}

#[test]
fn type_not_found() {
    assert_yaml_snapshot!(check_err(&[(
        "root.proto",
        "
        message Foo {
            optional NotFound foo = 1;
        }
    "
    )]));
}

#[test]
fn default_options() {
    let mut compiler = Compiler::with_file_resolver(TestFileResolver {
        files: &[("dep.proto", ""), ("root.proto", "import 'dep.proto';")],
    });

    compiler.open_file("root.proto").unwrap();

    let files = compiler.file_descriptor_set();
    assert_eq!(
        files,
        FileDescriptorSet {
            file: vec![FileDescriptorProto {
                name: Some("root.proto".to_owned()),
                dependency: vec!["dep.proto".to_owned()],
                ..Default::default()
            },],
        }
    );

    let encoded = compiler.encode_file_descriptor_set();
    assert_eq!(
        FileDescriptorSet::decode(encoded.as_slice()).unwrap(),
        files
    );
}

#[test]
fn include_imports() {
    let mut compiler = Compiler::with_file_resolver(TestFileResolver {
        files: &[("dep.proto", ""), ("root.proto", "import 'dep.proto';")],
    });

    compiler.include_imports(true);
    compiler.open_file("root.proto").unwrap();

    let files = compiler.file_descriptor_set();
    assert_eq!(
        files,
        FileDescriptorSet {
            file: vec![
                FileDescriptorProto {
                    name: Some("dep.proto".to_owned()),
                    ..Default::default()
                },
                FileDescriptorProto {
                    name: Some("root.proto".to_owned()),
                    dependency: vec!["dep.proto".to_owned()],
                    ..Default::default()
                },
            ],
        }
    );

    let encoded = compiler.encode_file_descriptor_set();
    assert_eq!(
        FileDescriptorSet::decode(encoded.as_slice()).unwrap(),
        files
    );
}

#[test]
fn include_source_info() {
    let mut compiler = Compiler::with_file_resolver(TestFileResolver {
        files: &[("dep.proto", ""), ("root.proto", "import 'dep.proto';")],
    });

    compiler.include_source_info(true);
    compiler.open_file("root.proto").unwrap();

    let files = compiler.file_descriptor_set();
    assert_eq!(
        files,
        FileDescriptorSet {
            file: vec![FileDescriptorProto {
                name: Some("root.proto".to_owned()),
                dependency: vec!["dep.proto".to_owned()],
                source_code_info: Some(SourceCodeInfo {
                    location: vec![
                        Location {
                            path: vec![],
                            span: vec![0, 0, 19],
                            ..Default::default()
                        },
                        Location {
                            path: vec![3, 0],
                            span: vec![0, 0, 19],
                            ..Default::default()
                        }
                    ]
                }),
                ..Default::default()
            },],
        }
    );

    let encoded = compiler.encode_file_descriptor_set();
    assert_eq!(
        FileDescriptorSet::decode(encoded.as_slice()).unwrap(),
        files
    );
}

#[test]
fn pass_through_extension_options() {
    let mut resolver = ChainFileResolver::new();
    resolver.add(TestFileResolver {
        files: &[(
            "root.proto",
            "
            import 'google/protobuf/descriptor.proto';

            extend google.protobuf.FileOptions {
                optional int32 ext = 1001;
            }

            option (ext) = 1;
        ",
        )],
    });
    resolver.add(GoogleFileResolver::new());

    let mut compiler = Compiler::with_file_resolver(resolver);
    compiler.include_imports(true);
    compiler.open_file("root.proto").unwrap();

    let dyn_set = DescriptorPool::decode(compiler.encode_file_descriptor_set().as_slice()).unwrap();
    let ext = dyn_set.get_extension_by_name("ext").unwrap();
    assert_eq!(
        dyn_set
            .get_file_by_name("root.proto")
            .unwrap()
            .options()
            .get_extension(&ext)
            .as_ref(),
        &Value::I32(1)
    );

    let roundtripped_resolver =
        DescriptorSetFileResolver::decode(compiler.encode_file_descriptor_set().as_slice())
            .unwrap();
    let mut roundtripped_compiler = Compiler::with_file_resolver(roundtripped_resolver);
    roundtripped_compiler.include_imports(true);

    roundtripped_compiler.open_file("root.proto").unwrap();
    let roundtripped_dyn_set = DescriptorPool::decode(
        roundtripped_compiler
            .encode_file_descriptor_set()
            .as_slice(),
    )
    .unwrap();
    let roundtripped_ext = roundtripped_dyn_set.get_extension_by_name("ext").unwrap();
    assert_eq!(
        roundtripped_dyn_set
            .get_file_by_name("root.proto")
            .unwrap()
            .options()
            .get_extension(&roundtripped_ext)
            .as_ref(),
        &Value::I32(1)
    );
}
