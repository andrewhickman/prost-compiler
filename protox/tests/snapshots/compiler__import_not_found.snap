---
source: protox/tests/compiler.rs
expression: "check_err(&[(\"root.proto\", \"import 'notfound.proto';\")])"
---
causes: []
labels: []
message: "import 'notfound.proto' not found"
related: []
severity: error

